### PR TITLE
[spirv] Add CL option to disable SPIR-V legalization

### PR DESCRIPTION
**Pull request by antiagainst**

### CODE REVIEW

**Comment by antiagainst** [comment]
Hey @marcelolr, this PR adds a new command-line option: `-fspirv-disable-legal`, so it touches `HLOptions.h`, `HLOptions.cpp`, `HLOptions.td`, `dxcompilerobj.cpp`.

**Comment by antiagainst** [comment]
@ehsannas: I feel that is kinda too long. Just like we typically use opt to mean optimization, I think it's fine to use legal to mean legalization.

**Comment by marcelolr** [comment]
This doesn't look unreasonable and I'm happy to go with it - I'm wondering what's the use case? At first I thought it might be similar to the existing /Vd to disable validation, but this is intentionally codegen'ing something different.

**Comment by antiagainst** [comment]
@marcelolr: This is actually a very interesting one. :)

We know that to fully determine whether HLSL source code is legal or not, a few optimization passes must be run. This is also true for SPIR-V. Some code, if translated literally, is not valid for Vulkan consumption:

```hlsl
struct CombinedTexSampler {
  Texture2D tex;
  SamplerState sampler;
}

float4 doSample(CombinedTexSampler comb) {
  return comb.tex.Sample(tex.sampler, ...);
}

Texture2D gTex;
SamplerState gSampler;

float4 main() : SV_Target {
  CombinedTexSampler comb;
  comb.tex = gTex;
  comb.sampler = gSampler;
  return doSample(comb);
}
```

The above, if translated literally, will generate a `OpTypeStruct` with two fields: `OpTypeImage` and `OpTypeSampler`. That is illegal SPIR-V since both `OpTypeImage` and `OpTypeSampler` are opaque types whose size and memory layout is unknown. And Vulkan also does not support writing to variables of opaque types.

In order to generate legal SPIR-V for the above code, function inlining and a few other optimization passes need to be run. We call these passes as legalization passes here since now their purpose are to make the generated SPIR-V legal. Since we do our CodeGen directly from AST, we cannot share these optimizations from the DXIL side. So we use the optimization passes in SPIRV-Tools.

Developers should not turn off legalization passes since that may generate illegal SPIR-V but we sometimes want to investigate the raw output from SPIR-V CodeGen, without the modification from SPIRV-Tools.

So basically, this command line is for internal debugging and testing purpose. That's why it is purposely kept undocumented.

**Comment by antiagainst** [comment]
`/Vd` will also be used later in SPIR-V CodeGen to disable validation. Right now we just don't run validation at all except in testing.

**Comment by marcelolr** [comment]
Ah, this is similar to what we produce with /fcgl then. Would it be useful to piggy back on that? That writes out an unprocessed output from the AST (as LLVM in the DXIL case), with the idea of debugging and/or experimenting with different optimization pipelines (some of which are required to produce valid DXIL).

**Comment by antiagainst** [comment]
Oh, sweet! That seems exactly what we need. Just the name is not that obvious so I didn't spot it previously. I've adjusted to use fcgi now. Expertise helps! Much appreciated! :)
